feat(aws-apigateway-kinesisstreams): New aws-apigateway-kinesisstreams pattern implementation

dscpinheiro opened the pull request:
*Issue #, if available:* #51

*Description of changes:* This PR introduces a new pattern, where data can be ingested to a Kinesis Data Streams via API Gateway. 

API Gateway acts as a proxy, providing a layer of abstraction from the streaming storage. This enables customers to implement custom authentication approaches for data producers, control quotas for specific producers, and change the target Kinesis stream.

The pattern only includes the _PutRecord_ and _PutRecords_ actions in Kinesis, as there are other alternatives (such as the [aws-kinesisstreams-lambda](https://docs.aws.amazon.com/solutions/latest/constructs/aws-kinesisstreams-lambda.html) construct) for reading data from a stream.

In addition to the pattern, other changes also included in this PR are:
- Update `kinesis-streams-helper` to allow using an `existingStreamObj` 
- Update `apigateway-helper` to specify optional `RequestValidator` and `Model` when adding a proxy method to an API resource.

The request templates used for the API methods follow the structures described in this documentation page (https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-kinesis.html).
- PutRecord:
```
{
    "StreamName": "$input.params('stream-name')",
    "Data": "$util.base64Encode($input.json('$.data'))",
    "PartitionKey": "$input.path('$.partitionKey')"
}
```

- PutRecords:
```
{
    "StreamName": "$input.params('stream-name')",
    "Records": [
        #foreach($elem in $input.path('$.records'))
          {
            "Data": "$util.base64Encode($elem.data)",
            "PartitionKey": "$elem.partitionKey"
          }#if($foreach.hasNext),#end
        #end
    ]
}
```

Since these templates require a specific format, the `aws-apigateway-kinesisstreams` construct will also create the corresponding [models](https://docs.aws.amazon.com/apigateway/latest/developerguide/rest-api-data-transformations.html#models-mappings-models), which will then be used for [request validation](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-method-request-validation.html). Without the models and validation, invalid request payloads would return a 200 response, even though the integration could not be invoked successfully:

```
Request: POST /record
Request body: { "foo": "bar" }

Response status: 200
Response body:
{
  "__type": "ValidationException",
  "message": "1 validation error detected: Value '' at 'partitionKey' failed to satisfy constraint: Member must have length greater than or equal to 1"
}
```

Model creation can be disabled by setting the `createRequestModels` property of `ApiGatewayToKinesisStreamsProps` to false, but it's set to true by default.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

dscpinheiro commented on source/patterns/@aws-solutions-constructs/core/lib/apigateway-helper.ts:
```diff
@@ -191,7 +191,9 @@ export interface AddProxyMethodToApiResourceInputParams {
-    readonly contentType?: string
+    readonly contentType?: string,
+    readonly requestValidator?: api.RequestValidator,
+    readonly requestModel?: { [contentType: string]: api.IModel; }
```
This is an odd type definition, but it matches the one from [MethodOptions.requestModels](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-apigateway.MethodOptions.html):

```typescript
  /**
   * The models which describe data structure of request payload.
   * Specify `requestModels` as key-value pairs, with a content type
   * (e.g. `'application/json'`) as the key and an API Gateway Model as the value.
   */
  readonly requestModels?: { [param: string]: IModel };
```

dscpinheiro commented on source/patterns/@aws-solutions-constructs/aws-apigateway-kinesisstreams/README.md:
```diff
@@ -0,0 +1,105 @@
+![Architecture Diagram](architecture.png)
+
+***
+&copy; Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
I noticed the copyright year on the README is 2020, but all header files say 2019. Is that a problem?

aws-solutions-constructs-team commented:
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-rp7jMDIK1wBN
* Commit ID: 0a5f1eb04d5e8b04ab06718075a7065640d77085
* Result: FAILED
* [Build Logs](https://nggetq65qd.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=48d02deb-e8d1-4b92-b5bc-a1c413e4f388%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dscpinheiro commented:
I'm looking into the build error. It's failing because of this section on `integ.apigateway-kinesis-overwrite.ts`:

```typescript
    kinesisStreamProps: {
        shardCount: 1,
        streamName: 'my-stream'
    },
```

cfn_nag then warns that _"Resource found with an explicit name, this disallows updates that require replacement of this resource"_.

I'll remove the `streamName` property.

aws-solutions-constructs-team commented:
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-rp7jMDIK1wBN
* Commit ID: 2dcbae01c9a5756e430ef6736ef0202960841651
* Result: SUCCEEDED
* [Build Logs](https://nggetq65qd.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d706a99b-169f-4984-9e04-ae99c5abca1e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

hnishar commented on source/patterns/@aws-solutions-constructs/core/test/apigateway-helper.test.ts:
```diff
@@ -286,4 +286,41 @@ test('Test addMethodToApiResource with action', () => {
+  expect(stack).toHaveResource('AWS::ApiGateway::RequestValidator', {
+    Name: "default-validator",
+    ValidateRequestBody: true
+  });
```
Add the assert for `RequestModels` to be empty ?

hnishar reviewed (changes requested):
Overall looks good, apart from minor comments, main feedback is around how to handle override for `RequestTemplates` and `RequestModels`

aws-solutions-constructs-team commented:
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-rp7jMDIK1wBN
* Commit ID: a1552c342bbbe224062388ea8583ece2add1f860
* Result: SUCCEEDED
* [Build Logs](https://nggetq65qd.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=161051fe-b8dd-42a9-85be-2ccad7e7986e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dscpinheiro commented:
Thanks for the feedback @hnishar, I've addressed your comments in my latest commit (https://github.com/awslabs/aws-solutions-constructs/pull/56/commits/e3b3f8f9ce15717d6a2fe7f85b1608ed0999936c).

dscpinheiro commented on source/patterns/@aws-solutions-constructs/aws-apigateway-kinesisstreams/test/apigateway-kinesis.test.ts:
```diff
@@ -0,0 +1,108 @@
+        ShardCount: 5,
+        RetentionPeriodHours: 96
+    });
+});
```
I was testing the snapshot and the properties in a single function. I split them into two separate tests ('Test minimal deployment snapshot' and 'Test construct properties').

aws-solutions-constructs-team commented:
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-rp7jMDIK1wBN
* Commit ID: e3b3f8f9ce15717d6a2fe7f85b1608ed0999936c
* Result: SUCCEEDED
* [Build Logs](https://nggetq65qd.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8d965b8b-2b48-4c76-a385-18a4c722fbdb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dscpinheiro commented:
I found a bug with my latest changes: the _addModel_ method was not being invoked for user provided models. I've converted the PR to draft while I fix it.

dscpinheiro commented on source/patterns/@aws-solutions-constructs/aws-apigateway-kinesisstreams/README.md:
```diff
@@ -55,9 +55,9 @@ _Parameters_
 |apiGatewayProps?|[`api.RestApiProps`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-apigateway.RestApiProps.html)|Optional user-provided props to override the default props for the API Gateway.|
 |putRecordRequestTemplate?|`string`|API Gateway request template for the PutRecord action. If not provided, a default one will be used.|
-|putRecordRequestModel?|[`api.Model`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-apigateway.Model.html)|API Gateway request model for the PutRecord action. If not provided, a default one will be created.|
+|putRecordRequestModel?|[`api.ModelOptions`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-apigateway.ModelOptions.html)|API Gateway request model for the PutRecord action. If not provided, a default one will be created.|
```
The reason I changed this property to `ModelOptions` is because the _RestApi.addModel_ method takes a `ModelOptions` interface as a parameter (https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-apigateway.RestApi.html#add-wbr-modelid-props)

aws-solutions-constructs-team commented:
### AWS CodeBuild CI Report

* CodeBuild project: codebuildgithubautobuildPro-rp7jMDIK1wBN
* Commit ID: 8963808708d6844258696ac4191a4f9c6cee39af
* Result: SUCCEEDED
* [Build Logs](https://nggetq65qd.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9893fca5-a0c3-431c-b965-99ef37708f54%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

hnishar reviewed (approved):
Ship it! Thanks for you contribution !
